crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+) -> String {
+    let prelude_end = {
+        let first_item_line_offset = locator.line_start(first_item_start_offset);
+        if first_item_line_offset == locator.line_start(dunder_all_offset) {
+            first_item_start_offset
+        } else {
+            first_item_line_offset
+        }
+    };
+    let prelude = locator.slice(TextRange::new(dunder_all_offset, prelude_end));
+    format!("{}{}", prelude.trim_end(), newline)
+}
+
+fn join_multiline_dunder_all_items(
+    sorted_items: &[DunderAllItem],
+    locator: &Locator,
+    item_indent: &str,
+    newline: &str,
+    needs_trailing_comma: bool,
+) -> String {
+    let max_index = sorted_items.len() - 1;
+
+    let mut new_dunder_all = String::new();
+    for (i, item) in sorted_items.iter().enumerate() {
+        let is_final_item = i == max_index;
+        for comment_range in &item.preceding_comment_ranges {
+            new_dunder_all.push_str(item_indent);
+            new_dunder_all.push_str(locator.slice(comment_range));
+            new_dunder_all.push_str(newline);
+        }
+        new_dunder_all.push_str(item_indent);
+        new_dunder_all.push_str(locator.slice(item.element_range));
+        if !is_final_item || needs_trailing_comma {
+            new_dunder_all.push(',');
+        }
+        if let Some(trailing_comments) = item.end_of_line_comments {
+            new_dunder_all.push_str(locator.slice(trailing_comments));
+        }
+        if !is_final_item {
+            new_dunder_all.push_str(newline);
+        }
+    }
+    new_dunder_all
+}
+
+fn multiline_dunder_all_postlude<'a>(
+    last_item_end_offset: TextSize,
+    newline: &str,
+    leading_indent: &str,
+    item_indent: &str,
+    dunder_all_range_end: TextSize,
+    locator: &'a Locator,
+) -> Cow<'a, str> {
+    let postlude_start = {
+        let last_item_line_offset = locator.line_end(last_item_end_offset);
+        if last_item_line_offset == locator.line_end(dunder_all_range_end) {
+            last_item_end_offset
+        } else {
+            last_item_line_offset
+        }
+    };
+    let postlude = locator.slice(TextRange::new(postlude_start, dunder_all_range_end));
+    if !postlude.starts_with(newline) {
+        return Cow::Borrowed(postlude);
+    }
+    if TextSize::of(leading_indentation(postlude.trim_start_matches(newline)))
+        <= TextSize::of(item_indent)
+    {
+        return Cow::Borrowed(postlude);
+    }
+    let trimmed_postlude = postlude.trim_start();
+    if trimmed_postlude.starts_with(']') || trimmed_postlude.starts_with(')') {
+        return Cow::Owned(format!("{newline}{leading_indent}{trimmed_postlude}"));
+    }
+    Cow::Borrowed(postlude)
+}
+
+fn sort_single_line_dunder_all(
+    elts: &[ast::Expr],
+    elements: &[&str],
+    kind: &DunderAllKind,
+    locator: &Locator,
+) -> String {
+    let mut element_pairs = elts.iter().zip(elements).collect_vec();
+    element_pairs.sort_by_cached_key(|(_, elem)| AllItemSortKey::from(**elem));
```

Comment:
Nit: Using `sort_by` is probably sufficient now that we can avoid allocating in `AllItemSortKey`